src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/poptimizer/simple/CardinalityBasedGreedyJoinPlanOptimizerImpl.java

```diff
@@ -0,0 +1,291 @@
+                        fms.add( fm );
+
+                        index++;
+                    }
+
+                    planWithStatistics.addCandidate( sumCard );
+                    planWithStatistics.addNumOfAccess( sumAccess );
+                    planWithStatistics.addFederationMembers( fms );
+                }
+                else
+                    throw new IllegalArgumentException("Unsupported type of subquery in source assignment (" + pop.getClass().getName() + ")");
+
+                subPlansWithStatistics.add( planWithStatistics );
+            }
+
+            return subPlansWithStatistics;
+        }
+
+        /**
+         * Compares all available subplans in terms of their cardinality estimation
+         * and returns the one with the lowest estimated cardinality.
+         */
+        protected PhysicalPlanWithStatistics chooseNextSubPlanBasedOnCard( final List<PhysicalPlanWithStatistics> subPlansWithStatistics ) {
+            PhysicalPlanWithStatistics bestPlanWithCandidate = subPlansWithStatistics.get(0);
+            int lowestCost = bestPlanWithCandidate.getCandidate();
+
+            for ( int i = 1; i < subPlansWithStatistics.size(); i++ ) {
+                PhysicalPlanWithStatistics current = subPlansWithStatistics.get(i);
+                if ( current.getCandidate() < lowestCost ) {
+                    lowestCost = current.getCandidate();
+                    bestPlanWithCandidate = current;
+                }
+            }
+
+            return bestPlanWithCandidate;
+        }
+
+        /**
+         * The physical algorithm is determined based on the estimated number of requests to execute the join
+         * - SHJ: Card(T1)/PageSize + Card(T2)/PageSize
+         * - BJ: Card(T1)/PageSize + Card(T1)/blockSize
+         * For the comparison, only keeping the second element is enough.
+         */
+        protected PhysicalPlanWithStatistics decidePhysicalAlgorithm( final PhysicalPlanWithStatistics currentPlan, final PhysicalPlanWithStatistics nextPlan ) {
+            final double accNumSHJ = nextPlan.getNumOfAccess();
+
+            final List<FederationMember> fms = nextPlan.getFederationMembers();
+            double accNumBJ = 0;
+            for ( final FederationMember fm: fms ) {
+                double blockSize = setBlockSize(fm);
+                accNumBJ += currentPlan.getCandidate() / blockSize;
+            }
+
+            PhysicalPlan newPlan;
+            if ( accNumSHJ <= accNumBJ ) {
+                newPlan = PhysicalPlanFactory.createPlanWithJoin(currentPlan.plan, nextPlan.plan);
+            }
+            else {
+                newPlan = PhysicalPlanFactory.createPlanWithDefaultUnaryOpIfPossible(currentPlan.plan, nextPlan.plan);
+            }
+
+            final PhysicalPlanWithStatistics newPlanWithStatistics = new PhysicalPlanWithStatistics( newPlan );
+
+            final int joinCardinality = Math.min( currentPlan.getCandidate(), nextPlan.getCandidate() );
```

Comment:
Can you please add a comment here that refers to the corresponding part/formula of the paper, where they say that they use min for this.